### PR TITLE
sql: disambiguate ARRAY[] type

### LINKED ISSUE #38293: sql: inserting with default values for both INT ARRAY and STRING ARRAY columns fails
**Describe the problem**

When adding rows using default values to tables with `INT[] NOT NULL` and `STRING[] NOT NULL` columns (default=empty array) the insert fails.

The error message is either:

- `pq: type of array contents int doesn't match column type STRING`
- or `pq: type of array contents string doesn't match column type INT`.

depending which column is defined first. So the second column will produce the error for some reason.

If there are two array columns of same type this issue doesn't seem to occur.

**To Reproduce**

Case 1:
```
CREATE TABLE IF NOT EXISTS bugtest1 (
    id SERIAL NOT NULL PRIMARY KEY,
    arr1 STRING(100) ARRAY NOT NULL default ARRAY[],
    arr2 INT ARRAY NOT NULL default ARRAY[]
);
insert into bugtest1 (id) values (1);
```
Result:
`pq: type of array contents string doesn't match column type INT`

Case 2:
```
CREATE TABLE IF NOT EXISTS bugtest2 (
    id SERIAL NOT NULL PRIMARY KEY,
    arr2 INT ARRAY NOT NULL default ARRAY[],
    arr1 STRING(100) ARRAY NOT NULL default ARRAY[]
);
insert into bugtest2 (id) values (1);
```
Result:
`pq: type of array contents int doesn't match column type STRING`

**Expected behavior**

DB should use the default value provided and insert to row successfully. In previous version I used (2.x.x) this or at least comparable example used to work.

**Additional data / screenshots**

**Environment:**

 - CockroachDB version: CockroachDB CCL v19.1.2 (x86_64-apple-darwin18.6.0, built 2019/06/17 20:21:54, go1.12. (installed using brew)
 - Server OS: Mac OS X 10.14.4
 - Client app: `cockroach sql`

**Additional context**

Running a migration tools to get the database up to date now fails because earlier migration would assume this syntax works (and it worked for a long time). Default values for array types seem to be very broken in this version.

This issue could be somewhat related: [Issue 23429](https://github.com/cockroachdb/cockroach/issues/23429)

### PR DESCRIPTION
**Pull request by RaduBerinde**
The optimizer uses `AsStringWithFlags` with `FmtCheckEquivalence` to
deduplicate expressions. If we have two empty array expressions that
are resolved to different types, the string is just `ARRAY[]` and we
incorrectly assume the expressions are the same. The fix is to add a
type annotation in this case.

Fixes #38293.

Release note (bug fix): Fixed an incorrect type mismatch error when
empty array values are used as default values (and potentially in
other contexts).

### CODE REVIEW

File: pkg/sql/sem/tree/expr.go [hunk 1]
```diff
@@ -898,6 +898,10 @@ func (node *Array) Format(ctx *FmtCtx) {
 	ctx.WriteString("ARRAY[")
 	ctx.FormatNode(&node.Exprs)
 	ctx.WriteByte(']')
+	if len(node.Exprs) == 0 && ctx.HasFlags(FmtParsable) && node.typ != nil {
```
Comment (jordanlewis) [inline pkg/sql/sem/tree/expr.go]: Ah... one of these problems? Man wouldn't it be nice to come up with a proto-based serialization format soon.

Is this going to be a problem for `DArray` too? In that case we can fix it by adjusting the `AmbiguousFormat` method. I wonder if the caveats currently in that method need to apply to this too...

**Review by jordanlewis — APPROVED** [review]

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38294)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/sem/tree/expr.go, line 901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38294#-Lhkp3xZpkrcwrxAsXD9-r1-901:-LhktO178t4CnUKGQvBG:b-fqrz7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0536d8187fff330ba3d7693f0ee066c6d5602d2f/pkg/sql/sem/tree/expr.go#L901)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ah... one of these problems? Man wouldn't it be nice to come up with a proto-based serialization format soon.

Is this going to be a problem for `DArray` too? In that case we can fix it by adjusting the `AmbiguousFormat` method. I wonder if the caveats currently in that method need to apply to this too...
</blockquote></details>

DArray looks fine (`AmbiguousFormat` would return true in this case). What caveats are you talking about?

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38294)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/sem/tree/expr.go, line 901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38294#-Lhkp3xZpkrcwrxAsXD9-r1-901:-Lhkw7J52jQQ1CymwarM:bg6c603) ([raw file](https://github.com/cockroachdb/cockroach/blob/0536d8187fff330ba3d7693f0ee066c6d5602d2f/pkg/sql/sem/tree/expr.go#L901)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

DArray looks fine (`AmbiguousFormat` would return true in this case). What caveats are you talking about?
</blockquote></details>

if the array only had nulls in it but had a type. so node.Exprs is nonempty but node.typ is not nil. in that case you still need a type annotation.

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/sem/tree/expr.go, line 901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38294#-Lhkp3xZpkrcwrxAsXD9-r1-901:-Lhkx89_72hGF5-CIEvQ:b5457hb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0536d8187fff330ba3d7693f0ee066c6d5602d2f/pkg/sql/sem/tree/expr.go#L901)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

if the array only had nulls in it but had a type. so node.Exprs is nonempty but node.typ is not nil. in that case you still need a type annotation.
</blockquote></details>

Oh. My thought was that the `NULL`s themselves would be typed, but alas that's not the case.. I could check for NULLs but that seems fragile here.. What do you think about always adding the annotation for `ARRAY`?

**Comment by jordanlewis** [comment]
> Oh. My thought was that the NULLs themselves would be typed, but alas that's not the case.. I could check for NULLs but that seems fragile here.. What do you think about always adding the annotation for ARRAY?

That sounds fine to me. I don't think these get serialized that much.

**Review by RaduBerinde — COMMENTED** [review]
Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38294#-:-Lhl2ogu1idOCN9mhhl9:b-896fix)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

**Comment by jordanlewis** [comment]
LGTM

**Pull request merged**